File: go/internal/scanner/scanner.go
// Package scanner is the native Go implementation of envdoctor's core:
// reconcile environment variables used in Go source against those defined in
// .env files. Local-first — no network, values never printed.
package scanner

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// Origin is where a variable was seen.
type Origin struct {
	File string
	Line int
}

// Finding is one reported issue.
type Finding struct {
	Rule     string
	Severity string // "error" | "warning"
	Name     string
	Message  string
	Origin   Origin
}

// Result holds all findings from a scan.
type Result struct {
	Findings []Finding
}

// Errors returns only error-severity findings.
func (r Result) Errors() []Finding { return r.filter("error") }

// Warnings returns only warning-severity findings.
func (r Result) Warnings() []Finding { return r.filter("warning") }

func (r Result) filter(sev string) []Finding {
	var out []Finding
	for _, f := range r.Findings {
		if f.Severity == sev {
			out = append(out, f)
		}
	}
	return out
}

var usagePatterns = []*regexp.Regexp{
	regexp.MustCompile(`\bos\.Getenv\(\s*"([A-Za-z_]\w*)"`),
	regexp.MustCompile(`\bos\.LookupEnv\(\s*"([A-Za-z_]\w*)"`),
}

var (
	lineComment  = regexp.MustCompile(`(?m)//[^\n]*`)
	blockComment = regexp.MustCompile(`(?s)/\*.*?\*/`)
	envLine      = regexp.MustCompile(`^\s*(?:export\s+)?([A-Za-z_]\w*)\s*=`)
)

// publicPrefixes are client-exposed env prefixes (case-sensitive, exact).
var publicPrefixes = []string{
	"NEXT_PUBLIC_",
	"VITE_",
	"REACT_APP_",
	"EXPO_PUBLIC_",
	"GATSBY_",
	"NUXT_PUBLIC_",
	"VUE_APP_",
	"PUBLIC_",
}

// secretName matches secret-looking variable names (case-insensitive).
var secretName = regexp.MustCompile(`(?i)SECRET|TOKEN|PASSWORD|PASSWD|PRIVATE|CREDENTIAL|API_?KEY|ACCESS_?KEY|AUTH`)

func hasPublicPrefix(name string) bool {
	for _, p := range publicPrefixes {
		if strings.HasPrefix(name, p) {
			return true
		}
	}
	return false
}

// blankMatch replaces every non-newline rune with a space to preserve offsets.
func blankMatch(s string) string {
	var b strings.Builder
	for _, r := range s {
		if r == '\n' {
			b.WriteRune('\n')
		} else {
			b.WriteByte(' ')
		}
	}
	return b.String()
}

func stripNoise(code string) string {
	code = blockComment.ReplaceAllStringFunc(code, blankMatch)
	code = lineComment.ReplaceAllStringFunc(code, blankMatch)
	return code
}

// ScanSource returns variable name -> first origin for env usage in Go source.
func ScanSource(path, content string) map[string]Origin {
	text := stripNoise(content)
	used := map[string]Origin{}
	for _, re := range usagePatterns {
		for _, m := range re.FindAllStringSubmatchIndex(text, -1) {
			name := text[m[2]:m[3]]
			if _, ok := used[name]; ok {
				continue
			}
			line := strings.Count(text[:m[0]], "\n") + 1
			used[name] = Origin{File: path, Line: line}
		}
	}
	return used
}

// ParseEnv returns variable name -> all origins (in file order) for the
// definitions in a dotenv file. Collecting every occurrence lets callers both
// reconcile against the first origin and detect in-file duplicates.
func ParseEnv(path, content string) map[string][]Origin {
	defined := map[string][]Origin{}
	for i, raw := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(raw)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		if m := envLine.FindStringSubmatch(raw); m != nil {
			defined[m[1]] = append(defined[m[1]], Origin{File: path, Line: i + 1})
		}
	}
	return defined
}

func skipDir(name string) bool {
	switch name {
	case ".git", "vendor", "node_modules":
		return true
	}
	return false
}

// Scan reconciles .env definitions against .go source usage under root.
func Scan(root string) (Result, error) {
	defined := map[string]Origin{}
	used := map[string]Origin{}
	var duplicates []Finding

	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			if skipDir(info.Name()) {
				return filepath.SkipDir
			}
			return nil
		}
		base := info.Name()
		isEnv := base == ".env" || (strings.HasPrefix(base, ".env.") && !strings.HasSuffix(base, ".example"))
		isGo := strings.HasSuffix(base, ".go")
		if !isEnv && !isGo {
			return nil
		}
		data, readErr := os.ReadFile(path)
		if readErr != nil {
			return readErr
		}
		rel, _ := filepath.Rel(root, path)
		if isEnv {
			for k, origins := range ParseEnv(rel, string(data)) {
				// First occurrence (across all files) is the definition.
				if _, ok := defined[k]; !ok {
					defined[k] = origins[0]
				}
				// Duplicate within a single file: 2+ occurrences of a key.
				if len(origins) >= 2 {
					lines := make([]string, len(origins))
					for i, o := range origins {
						lines[i] = strconv.Itoa(o.Line)
					}
					duplicates = append(duplicates, Finding{
						Rule: "duplicates", Severity: "error", Name: k,
						Message: fmt.Sprintf("defined %d times in the same file (lines %s)",
							len(origins), strings.Join(lines, ", ")),
						Origin: origins[0],
					})
				}
			}
		} else {
			for k, v := range ScanSource(rel, string(data)) {
				if _, ok := used[k]; !ok {
					used[k] = v
				}
			}
		}
		return nil
	})
	if err != nil {
		return Result{}, err
	}

	var res Result
	usedNames := sortedKeys(used)
	for _, name := range usedNames {
		if _, ok := defined[name]; !ok {
			res.Findings = append(res.Findings, Finding{
				Rule: "undefined-in-source", Severity: "error", Name: name,
				Message: "used in source code but not defined in any environment file",
				Origin:  used[name],
			})
		}
	}
	sort.Slice(duplicates, func(i, j int) bool { return duplicates[i].Name < duplicates[j].Name })
	res.Findings = append(res.Findings, duplicates...)
	for _, name := range sortedKeys(defined) {
		if hasPublicPrefix(name) && secretName.MatchString(name) {
			res.Findings = append(res.Findings, Finding{
				Rule: "public-prefix", Severity: "error", Name: name,
				Message: "secret-looking variable is exposed to client bundles via a public prefix",
				Origin:  defined[name],
			})
		}
	}
	for _, name := range sortedKeys(defined) {
		if _, ok := used[name]; !ok {
			res.Findings = append(res.Findings, Finding{
				Rule: "unused", Severity: "warning", Name: name,
				Message: "defined but never referenced in source",
				Origin:  defined[name],
			})
		}
	}
	return res, nil
}

func sortedKeys(m map[string]Origin) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

File: php/src/Scanner.php
<?php

declare(strict_types=1);

namespace Envdoctor;

/**
 * Core scanner: reconcile environment access in PHP source against .env
 * definitions. Local-first — no network, values never printed.
 */
final class Scanner
{
    /** @var string[] Each pattern captures the variable name in group 1. */
    private const USAGE_PATTERNS = [
        '/\bgetenv\(\s*["\']([A-Za-z_]\w*)["\']/',
        '/\$_ENV\[\s*["\']([A-Za-z_]\w*)["\']\s*\]/',
        '/\$_SERVER\[\s*["\']([A-Za-z_]\w*)["\']\s*\]/',
    ];

    private const ENV_LINE = '/^\s*(?:export\s+)?([A-Za-z_]\w*)\s*=/';

    /** @var string[] Public/client-bundle prefixes (case-sensitive). */
    private const PUBLIC_PREFIXES = [
        'NEXT_PUBLIC_', 'VITE_', 'REACT_APP_', 'EXPO_PUBLIC_',
        'GATSBY_', 'NUXT_PUBLIC_', 'VUE_APP_', 'PUBLIC_',
    ];

    private const SECRET_RE = '/SECRET|TOKEN|PASSWORD|PASSWD|PRIVATE|CREDENTIAL|API_?KEY|ACCESS_?KEY|AUTH/i';

    private static function blank(string $s): string
    {
        return preg_replace('/[^\n]/', ' ', $s);
    }

    private static function stripNoise(string $code): string
    {
        // Block comments, then line comments (// and #).
        $code = preg_replace_callback('/\/\*.*?\*\//s', fn($m) => self::blank($m[0]), $code);
        $code = preg_replace_callback('/(?:\/\/|#)[^\n]*/', fn($m) => self::blank($m[0]), $code);

        return $code;
    }

    /** @return array<string, array{file: string, line: int}> */
    public static function scanSource(string $path, string $content): array
    {
        $text = self::stripNoise($content);
        $used = [];
        foreach (self::USAGE_PATTERNS as $pattern) {
            if (preg_match_all($pattern, $text, $matches, PREG_OFFSET_CAPTURE)) {
                foreach ($matches[1] as $m) {
                    $name = $m[0];
                    if (isset($used[$name])) {
                        continue;
                    }
                    $line = substr_count(substr($text, 0, $m[1]), "\n") + 1;
                    $used[$name] = ['file' => $path, 'line' => $line];
                }
            }
        }

        return $used;
    }

    /**
     * Returns all occurrences per key: name => list of {file, line}, in order.
     *
     * @return array<string, array<int, array{file: string, line: int}>>
     */
    public static function parseEnv(string $path, string $content): array
    {
        $defined = [];
        foreach (explode("\n", $content) as $i => $raw) {
            $trimmed = trim($raw);
            if ($trimmed === '' || str_starts_with($trimmed, '#')) {
                continue;
            }
            if (preg_match(self::ENV_LINE, $raw, $m)) {
                $defined[$m[1]][] = ['file' => $path, 'line' => $i + 1];
            }
        }

        return $defined;
    }

    private static function isPublicPrefix(string $name): bool
    {
        $hasPrefix = false;
        foreach (self::PUBLIC_PREFIXES as $p) {
            if (str_starts_with($name, $p)) {
                $hasPrefix = true;
                break;
            }
        }

        return $hasPrefix && preg_match(self::SECRET_RE, $name) === 1;
    }

    /** @return string[] */
    private static function discoverFiles(string $root, string $suffix, bool $envMode): array
    {
        $out = [];
        $it = new \RecursiveIteratorIterator(
            new \RecursiveDirectoryIterator($root, \FilesystemIterator::SKIP_DOTS)
        );
        foreach ($it as $file) {
            $name = $file->getFilename();
            $skip = false;
            foreach (['.git', 'vendor', 'node_modules'] as $bad) {
                if (str_contains($file->getPathname(), DIRECTORY_SEPARATOR . $bad . DIRECTORY_SEPARATOR)) {
                    $skip = true;
                    break;
                }
            }
            if ($skip) {
                continue;
            }
            if ($envMode) {
                $isEnv = $name === '.env' || (str_starts_with($name, '.env.') && !str_ends_with($name, '.example'));
                if ($isEnv) {
                    $out[] = $file->getPathname();
                }
            } elseif (str_ends_with($name, $suffix)) {
                $out[] = $file->getPathname();
            }
        }
        sort($out);

        return $out;
    }

    /** @return Finding[] */
    public static function scan(string $root): array
    {
        $root = rtrim($root, DIRECTORY_SEPARATOR);
        $defined = [];
        $dupFindings = [];
        foreach (self::discoverFiles($root, '', true) as $f) {
            $rel = self::rel($root, $f);
            foreach (self::parseEnv($rel, (string) file_get_contents($f)) as $k => $origins) {
                if (count($origins) >= 2) {
                    $lines = array_map(fn($o) => $o['line'], $origins);
                    $dupFindings[] = new Finding(
                        'duplicates',
                        'error',
                        $k,
                        'defined ' . count($origins) . ' times in the same file (lines '
                            . implode(', ', $lines) . ')',
                        $origins[0]
                    );
                }
                // First occurrence (first file wins) counts as the definition.
                $defined[$k] ??= $origins[0];
            }
        }

        $used = [];
        foreach (self::discoverFiles($root, '.php', false) as $f) {
            foreach (self::scanSource(self::rel($root, $f), (string) file_get_contents($f)) as $k => $v) {
                $used[$k] ??= $v;
            }
        }

        $findings = [];
        $usedNames = array_keys($used);
        sort($usedNames);
        foreach ($usedNames as $name) {
            if (!isset($defined[$name])) {
                $findings[] = new Finding(
                    'undefined-in-source',
                    'error',
                    $name,
                    'used in source code but not defined in any environment file',
                    $used[$name]
                );
            }
        }
        usort($dupFindings, fn($a, $b) => strcmp($a->name, $b->name));
        foreach ($dupFindings as $finding) {
            $findings[] = $finding;
        }
        $definedNames = array_keys($defined);
        sort($definedNames);
        foreach ($definedNames as $name) {
            if (self::isPublicPrefix($name)) {
                $findings[] = new Finding(
                    'public-prefix',
                    'error',
                    $name,
                    'secret-looking variable is exposed to client bundles via a public prefix',
                    $defined[$name]
                );
            }
        }
        foreach ($definedNames as $name) {
            if (!isset($used[$name])) {
                $findings[] = new Finding(
                    'unused',
                    'warning',
                    $name,
                    'defined but never referenced in source',
                    $defined[$name]
                );
            }
        }

        return $findings;
    }

    private static function rel(string $root, string $path): string
    {
        return ltrim(substr($path, strlen($root)), DIRECTORY_SEPARATOR);
    }
}

/** One reported issue. */
final class Finding
{
    /** @param array{file: string, line: int} $origin */
    public function __construct(
        public string $rule,
        public string $severity,
        public string $name,
        public string $message,
        public array $origin,
    ) {
    }
}
